### DISCUSSION TITLE
When granting a role to another user, permissions don't seem to work

### DISCUSSION THREAD

**Discussion by thisisgoldman** [supabase/supabase#13896] [Questions]
Hi there,

We're hitting an error granting permissions in our Supabase database. Here is the scenario:

We create a new user in our Supabase database: `user_a` 

With that user, we create a new schema (i.e., `schema_a`) and a table (i.e., `table_a`) in the database. `user_a` is the owner of both the schema and table with permission to select, insert, and delete.

Initially, the new tables cannot be accessed by the Supabase client (i.e., the `anon` / `service_role` database users). So we grant `user_a`'s permissions to `anon`, etc:

```sql
grant user_a to anon;
```

We'd expect this grant to allow the `anon` user to also have usage on the new schema and tables - just as it works in Postgres. But this doesn't seem to be the case. Instead, we need to run a more explicit permission grant:

```sql
grant select on schema_a.table_a to anon;
```

The inability to pass permissions by granting a role to another role/user feels like a bug. Is this intentional, or are we missing a step?

**Comment by steve-chavez** [comment]
**Edit**: Disregard this answer. It was a local database that didn't have the `anon` configured as the Supabase platform.

> We'd expect this grant to allow the anon user to also have usage on the new schema and tables - just as it works in Postgres. But this doesn't seem to be the case.

I cannot reproduce this. It's working as intended:

```js
const { data, error } = await supabase
  .from('privileged_table')
  .select()

{"code":"42501","details":null,"hint":null,"message":"permission denied for table privileged_table"}
```

```sql
-- postgres is superuser here
grant postgres to anon;
```

```js
const { data, error } = await supabase
  .from('privileged_table')
  .select()

[]
```

Can you share the request plus the full error messsage?

---

Not sure if adding privileges to anon is the real intention here. You might want to look at this previous discussion for custom roles https://github.com/orgs/supabase/discussions/11948#discussioncomment-4802137

**Comment by steve-chavez** [comment]
> The inability to pass permissions by granting a role to another role/user feels like a bug. Is this intentional, or are we missing a step?

Note that this depends on whether the role is created with the `INHERIT` attribute. In the case of the `anon` role this is not true. You can run this on the supabase query editor to confirm:

```sql
select rolinherit from pg_roles where rolname = 'anon';

false
```

Basically the role was created like:

```sql
create role anon NOINHERIT;
```

And yes, this is intentional. So you cannot make the mistake of doing:

```sql
GRANT <highly_privileged> TO anon;
```

In one step and leak information to all unauthenticated users.

---

The `authenticated` and `service_role` are also NOINHERIT, but we could discuss lifting this restriction since these represent authenticated users.
